### PR TITLE
feat: allow immediate MacOS notifications

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Resolves https://github.com/electron/electron/issues/15789.

Refactors notification posting on MacOS to allow for some notifications to deliver immediately regardless of whether or not the subscribing app is active or not. 

Only `NSDistributedNotificationCenter` has access to the `deliverImmediate` selector, so I split out the `DoPostNotification` helper into the calling functions as it now make more sense for each to just call the methods themselves rather than call out to a helper. `deliverImmediate` was also added as an optional new param so as to maintain backwards compat.

/cc @ckerr

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

#### Release Notes

Notes: allow for MacOS notifications to be immediately delivered

### CODE REVIEW

File: docs/api/system-preferences.md [hunk 1]
```diff
@@ -59,10 +59,11 @@ Returns `Boolean` - Whether the system is in Dark Mode.
 
 Returns `Boolean` - Whether the Swipe between pages setting is on.
 
-### `systemPreferences.postNotification(event, userInfo)` _macOS_
+### `systemPreferences.postNotification(event, userInfo[, deliverImmediately])` _macOS_
 
 * `event` String
 * `userInfo` Object
+* `deliverImmediately` Boolean (optional) - `true` to post notifications immediately regardless of whether or not the subscribing app is active or not.
```
Comment (nornagon) [inline docs/api/system-preferences.md]:
```suggestion
* `deliverImmediately` Boolean (optional) - `true` to post notifications immediately regardless of whether or not the subscribing app is active or not. Defaults to `false`. See [NSDistributedNotificationCenter documentation](https://developer.apple.com/documentation/foundation/nsdistributednotificationcenter/1418360-postnotificationname) for details.
```

File: atom/browser/api/atom_api_system_preferences_mac.mm [hunk 2]
```diff
@@ -106,10 +106,18 @@ AVMediaType ParseMediaType(const std::string& media_type) {
 
 }  // namespace
 
-void SystemPreferences::PostNotification(
-    const std::string& name,
-    const base::DictionaryValue& user_info) {
-  DoPostNotification(name, user_info, kNSDistributedNotificationCenter);
```
Comment (nornagon) [inline atom/browser/api/atom_api_system_preferences_mac.mm]: can we delete the `NotificationCenterKind` enum now?

Comment (codebytere) [reply atom/browser/api/atom_api_system_preferences_mac.mm]: yes! will do

File: atom/browser/api/atom_api_system_preferences_mac.mm [hunk 3]
```diff
@@ -156,29 +171,6 @@ AVMediaType ParseMediaType(const std::string& media_type) {
...
-      center = [[NSWorkspace sharedWorkspace] notificationCenter];
-      break;
-    default:
-      break;
-  }
-  [center postNotificationName:base::SysUTF8ToNSString(name)
-                        object:nil
-                      userInfo:DictionaryValueToNSDictionary(user_info)];
-}
-
```
Comment (ckerr) [inline atom/browser/api/atom_api_system_preferences_mac.mm]: This function declaration should be removed from `atom_api_system_preferences.h` too

File: docs/api/system-preferences.md [hunk 4]
```diff
@@ -59,10 +59,11 @@ Returns `Boolean` - Whether the system is in Dark Mode.
 
 Returns `Boolean` - Whether the Swipe between pages setting is on.
 
-### `systemPreferences.postNotification(event, userInfo)` _macOS_
+### `systemPreferences.postNotification(event, userInfo[, deliverImmediately])` _macOS_
 
 * `event` String
 * `userInfo` Object
+* `deliverImmediately` Boolean (optional) - `true` to post notifications immediately regardless of whether or not the subscribing app is active or not. Defaults to `false`. See [NSDistributedNotificationCenter documentation](https://developer.apple.com/documentation/foundation/nsdistributednotificationcenter/1418360-postnotificationname) for details.
```
Comment (ckerr) [inline docs/api/system-preferences.md]: _(nit)_  "regardless of", "whether or not", and "active or not" all say the same thing.

```diff
- immediately regardless of whether or not the subscribing app is active or not
+ immediately even when the subscribing app is inactive
```

**Comment by nornagon** [comment]
Hey @codebytere, the release notes here are confusing because "notification" sounds like it means [Notification](https://electronjs.org/docs/api/notification). Can you edit @release-clerk's notes and clarify?

**Pull request merged**